pkg/storegateway/batch_series.go

```diff
@@ -0,0 +1,846 @@
+	chunksLimiter    ChunksLimiter           // Rate limiter for loading chunks.
+	seriesLimiter    SeriesLimiter           // Rate limiter for loading series.
+	skipChunks       bool                    // If true chunks are not loaded and minTime/maxTime are ignored.
+	minTime, maxTime int64                   // Series must have data in this time range to be returned (ignored if skipChunks=true).
+	loadAggregates   []storepb.Aggr          // List of aggregates to load when loading chunks.
+	logger           log.Logger
+}
+
+func unloadedBucketBatches(
+	ctx context.Context,
+	batchSize int,
+	indexr *bucketIndexReader, // Index reader for block.
+	blockID ulid.ULID,
+	matchers []*labels.Matcher, // Series matchers.
+	shard *sharding.ShardSelector, // Shard selector.
+	seriesHashCache *hashcache.BlockSeriesHashCache, // Block-specific series hash cache (used only if shard selector is specified).
+	chunksLimiter ChunksLimiter, // Rate limiter for loading chunks.
+	seriesLimiter SeriesLimiter, // Rate limiter for loading series.
+	skipChunks bool, // If true chunks are not loaded and minTime/maxTime are ignored.
+	minTime, maxTime int64, // Series must have data in this time range to be returned (ignored if skipChunks=true).
+	loadAggregates []storepb.Aggr, // List of aggregates to load when loading chunks.
+	logger log.Logger,
+) (unloadedBatchSet, error) {
+	if batchSize <= 0 {
+		return nil, errors.New("batch size must be a positive number")
+	}
+
+	stats := newSafeQueryStats()
+	ps, err := indexr.ExpandedPostings(ctx, matchers, stats)
+	if err != nil {
+		return nil, errors.Wrap(err, "expanded matching posting")
+	}
+
+	// We can't compute the series hash yet because we're still missing the series labels.
+	// However, if the hash is already in the cache, then we can remove all postings for series
+	// not belonging to the shard.
+	if shard != nil {
+		var unsafeStats queryStats
+		ps, unsafeStats = filterPostingsByCachedShardHash(ps, shard, seriesHashCache)
+		stats = stats.merge(&unsafeStats)
+	}
+
+	return &bucketBatchSet{
+		blockID:                 blockID,
+		batchSize:               batchSize,
+		currBatchPostingsOffset: -batchSize,
+		ctx:                     ctx,
+		postings:                ps,
+		indexr:                  indexr,
+		matchers:                matchers,
+		shard:                   shard,
+		seriesHasher:            cachedSeriesHasher{cache: seriesHashCache, stats: stats},
+		chunksLimiter:           chunksLimiter,
+		seriesLimiter:           seriesLimiter,
+		skipChunks:              skipChunks,
+		minTime:                 minTime,
+		maxTime:                 maxTime,
+		loadAggregates:          loadAggregates,
+		logger:                  logger,
+	}, nil
+}
+
+func (s *bucketBatchSet) Next() bool {
+	if s.currBatchPostingsOffset >= len(s.postings)-1 || s.err != nil {
+		return false
+	}
+	s.currentBatch = newBatch(s.batchSize)
+	return s.loadBatch()
+}
+
+func (s *bucketBatchSet) loadBatch() bool {
```

Comment:
Have you considered loading the actual chunk data from this function too? Without thinking too much, I don't think there's any practical different (in terms of resources consumption) from decoupling it but moving it here may make the code easier (especially with regards preloading that should preload the whole batch, as mentioned in another comment).

Reply:
If we load chunks in this function too we will keep at least 1 loaded batch for each block. The Series() RPC method returns the series one by one sorted by their labels. And in that case the order within blocks isn't really known, so it's not impossible that we return first batch of  series from the first block and then the second batch from the second block. In this case the batch from the second block will sit loaded in memory while we send the first one. 

This gets worse if the first block has a lot of batches. Then the batch from block2 will sit is memory while we send all those batches.

So if we load already merged and sorted series, we will hold a batch less time in memory.

Reply:
also gets worse the more blocks we have. We could also end up in a deadlock where we load many batches from many blocks but don't send most of them them because we need to load another batch from e.g. the first block. maybe not a deadlock, but a lot of batches would sit in memory

Reply:
am I missing something about your proposal?

Reply:
I see what you mean. You're right. It's good as you did 👍